contrib/Shopify/sarama/sarama.go

```diff
@@ -193,48 +198,47 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 		errors:        make(chan *sarama.ProducerError),
 	}
 	go func() {
-		type spanKey struct {
-			topic     string
-			partition int32
-			offset    int64
-		}
-		spans := make(map[spanKey]ddtrace.Span)
+		spans := make(map[uint64]ddtrace.Span)
 		defer close(wrapped.successes)
 		defer close(wrapped.errors)
 		for {
 			select {
 			case msg := <-wrapped.input:
-				key := spanKey{msg.Topic, msg.Partition, msg.Offset}
 				span := startProducerSpan(cfg, saramaConfig.Version, msg)
 				p.Input() <- msg
 				if saramaConfig.Producer.Return.Successes {
-					spans[key] = span
+					spanID := span.Context().SpanID()
+					spans[spanID] = span
 				} else {
 					// if returning successes isn't enabled, we just finish the
 					// span right away because there's no way to know when it will
 					// be done
-					finishProducerSpan(span, key.partition, key.offset, nil)
+					finishProducerSpan(span, nil, nil)
 				}
 			case msg, ok := <-p.Successes():
 				if !ok {
 					// producer was closed, so exit
 					return
 				}
-				key := spanKey{msg.Topic, msg.Partition, msg.Offset}
-				if span, ok := spans[key]; ok {
-					delete(spans, key)
-					finishProducerSpan(span, msg.Partition, msg.Offset, nil)
+				if spanctx, spanFound := getSpanContext(msg); spanFound {
+					spanID := spanctx.SpanID()
+					if span, ok := spans[spanID]; ok {
+						delete(spans, spanID)
+						finishProducerSpan(span, &producerSpanTag{Partition: msg.Partition, Offset: msg.Offset}, nil)
```

Comment:
I think the code would be a bit clearer by keeping the `finishProducerSpan` as-is, and just calling `span.Finish(...)` in cases where `partition` and `offset` don't need to be set. That avoids the need for the new `producerSpanTag` struct, the indirection, and therefore the `nil` check in `finishProducerSpan`.

Reply:
good call, done